### PR TITLE
IHub-179434 Initial implementation

### PR DESCRIPTION
**Pull request by P4radise**

### CODE REVIEW

File: settings.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "urlKMZ": "url",
+    "urlOneVizion": "url",
+    "loginOneVizion": "login",
```
Comment (asmoiseenko) [inline settings.json:4]: use access key and secret keys instead of username and password

File: kmz_to_kml.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
+from enum import Enum
```
Comment (asmoiseenko) [inline kmz_to_kml.py:1]: rename file to import_kml

File: README.md [hunk 3]
```diff
@@ -1 +1,10 @@
-# kml-import
\ No newline at end of file
+# kml_import
+
+The integration downloads the KMZ file, extracts data from it and writes them into a CSV file which starts the import of OneVizion
+
```
Comment (asmoiseenko) [inline README.md:4]: document currently supported KML struncture

File: README.md [hunk 4]
```diff
@@ -1 +1,10 @@
-# kml-import
\ No newline at end of file
+# kml_import
+
+The integration downloads the KMZ file, extracts data from it and writes them into a CSV file which starts the import of OneVizion
```
Comment (asmoiseenko) [inline README.md:3]: Downloads KMZ file from the specified location, extracts data to CSV file and loads it with the specified OneVizion import.

**Review by asmoiseenko — CHANGES_REQUESTED** [review]

File: kmz_to_kml.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+
+class File:
+
+    def __init__(self, url_fires, integration_log):
+        self.url_fires = url_fires
+        self.integration_log = integration_log
+
+    def get_csv(self):
+        kmz_file = self.download_kmz()
```
Comment (glebowadim) [inline kmz_to_kml.py:46]: kmz_content?

File: kmz_to_kml.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        self.integration_log.add_log(LogLevel.INFO.log_level_name, 'Integration has been completed')
+
+
+class File:
+
+    def __init__(self, url_fires, integration_log):
+        self.url_fires = url_fires
+        self.integration_log = integration_log
+
+    def get_csv(self):
```
Comment (glebowadim) [inline kmz_to_kml.py:45]: Let's separate to 2 classes. 1 class for working with KML(KMZ), 2 class for parsing and create csv file.

File: kmz_to_kml.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    def extract_kml(self, file):
+        kml = zipfile.ZipFile(file, 'r')
+        if len(kml.filelist) == 1:
+            kml_filename = kml.filelist[0].filename
+            kml.extract(kml_filename)
+            os.rename(kml_filename, FileNames.KML.value)
+            self.integration_log.add_log(LogLevel.INFO.log_level_name, 'KML file extracted')
+        else:
+            self.integration_log.add_log(LogLevel.ERROR.log_level_name, f'Failed extract kml file. Exception [kmz file is stored more than one file]')
+            raise Exception(f'Failed extract kml file. Exception [kmz file is stored more than one file]')
```
Comment (glebowadim) [inline kmz_to_kml.py:74]: ... Or less than one file. Is it so critical to receive more than 1 file?

File: kmz_to_kml.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        answer = requests.get(url)
+        if answer.ok:
+            self.integration_log.add_log(LogLevel.INFO.log_level_name, 'KMZ file downloaded')
+            return answer.content
+        else:
+            self.integration_log.add_log(LogLevel.ERROR.log_level_name, f'Failed download kmz. Exception [{answer.text}]')
+            raise Exception(f'Failed download kmz. Exception [{answer.text}]')
+
+    def save_kmz(self, file):
+        open(FileNames.KMZ.value, 'wb').write(file)
```
Comment (glebowadim) [inline kmz_to_kml.py:63]: use ```with```
https://realpython.com/python-with-statement/

File: kmz_to_kml.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+            if fire_type is None:
+                self.integration_log.add_log(LogLevel.ERROR.log_level_name, f'Failed parse kml. Exception [Field fire_type not found]')
+                raise Exception(f'Failed parse kml. Exception [Field fire_type not found]')
+
+            try:
+                fire_size = re.search(r'\d+', re.search(r'\d+\sacres|\d+acres', description).group()).group()
+            except Exception:
+                self.integration_log.add_log(LogLevel.ERROR.log_level_name, f'Failed parse kml. Exception [Field fire_size not found]')
+                raise Exception(f'Failed parse kml. Exception [Field fire_size not found]')
```
Comment (glebowadim) [inline kmz_to_kml.py:112]: Lost original exception

**Review by glebowadim — CHANGES_REQUESTED** [review]

File: kmz_to_kml.py (hunk 2 above)
Comment (P4radise) [reply kmz_to_kml.py]: Done

File: settings.json (hunk 1 above)
Comment (P4radise) [reply settings.json]: Done

File: README.md (hunk 4 above)
Comment (P4radise) [reply README.md]: Done

File: integration_log.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
+from enum import Enum
+from requests.auth import HTTPBasicAuth
+import json
+import requests
+
+
+class IntegrationLog(object):
```
[the remark this answers is not in the document]
Comment (P4radise) [reply integration_log.py]: Done

File: kmz_to_kml.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+            parse_list.append({CSVHeader.REG_ID.value:registration_id, CSVHeader.NAME.value:name, CSVHeader.DESCRIPTION.value:description, \
+                                CSVHeader.FIRE_SIZE.value:fire_size, CSVHeader.FIRE_TYPE.value:fire_type, CSVHeader.LATITUDE.value:coordinates[1], \
+                                    CSVHeader.LONGITUDE.value:coordinates[0]})
+
+        self.integration_log.add_log(LogLevel.INFO.log_level_name, 'KML file parsed')
+        return parse_list
+
+    def get_registration(self):
+        current_date = datetime.date.today().strftime('%m/%d/%y')
+        return f'Fires-{current_date}'
```
[the remark this answers is not in the document]
Comment (P4radise) [reply kmz_to_kml.py]: I found the date in the KML file, so now I take it from there and the constant is no longer needed

File: kmz_to_kml.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+                raise Exception(f'Import \"{self.import_name}\" not found')
+            else:
+                self.integration_log.add_log(LogLevel.INFO.log_level_name, f'Import \"{self.import_name}\" founded')
+        else:
+            self.integration_log.add_log(LogLevel.ERROR.log_level_name, f'Failed to receive import. Exception [{str(answer.text)}]')
+            raise Exception(f'Failed to receive import. Exception [{str(answer.text)}]')
+
+        return import_id
+
+    def start_import(self, import_id, file_name):
```
[the remark this answers is not in the document]
Comment (P4radise) [reply kmz_to_kml.py]: Done

File: kmz_to_kml.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+    REG_ID = 'RegistrationID'
+    NAME = 'Name'
+    DESCRIPTION = 'Description'
+    FIRE_SIZE = 'FireSize'
+    FIRE_TYPE = 'FireType'
+    LATITUDE = 'Latitude'
+    LONGITUDE = 'Longitude'
+
+
+class FileNames(Enum):
```
[the remark this answers is not in the document]
Comment (P4radise) [reply kmz_to_kml.py]: Done

File: kmz_to_kml.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+
+    def __init__(self, file, integration_import, integration_log):
+        self.file = file
+        self.integration_import = integration_import
+        self.integration_log = integration_log
+
+    def start_integration(self):
+        self.integration_log.add_log(LogLevel.INFO.log_level_name, 'Starting Integration')
+
+        try:
```
[the remark this answers is not in the document]
Comment (P4radise) [reply kmz_to_kml.py]: Done

File: kmz_to_kml.py (hunk 5 above)
Comment (P4radise) [reply kmz_to_kml.py]: Done

File: kmz_to_kml.py (hunk 6 above)
Comment (P4radise) [reply kmz_to_kml.py]: Done

File: kmz_to_kml.py (hunk 8 above)
Comment (P4radise) [reply kmz_to_kml.py]: Done

File: kmz_to_kml.py (hunk 7 above)
Comment (P4radise) [reply kmz_to_kml.py]: I added a loop, now the files will be searched, if several kml files are found or less than one, an Exception will be called

File: kmz_to_kml.py (hunk 9 above)
Comment (P4radise) [reply kmz_to_kml.py]: The original error would say that you cannot call the group function for an empty value, which means that no fire_size value was found.
If it is necessary, I can split this line and add some conditions.

Comment (asmoiseenko) [reply kmz_to_kml.py]: you know all exceptions which can be thrown? Never swallow original exception

File: import_kml.py [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+from enum import Enum
+import requests
+import zipfile
+import csv
+import re
+import os
+
+
+class Integration:
+    KMZ = 'fires.kmz'
```
Comment (asmoiseenko) [inline import_kml.py:13]: why fires? This is kml_import integration, it is not necessarily deal with fires related data. If you need to create temp files, use a different approach.

File: import_kml.py [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+        self.integration_log = integration_log
+
+    def get(self, kmz_file, kml_file):
+        kmz_content = self.download()
+        self.save(kmz_content, kmz_file)
+        self.extract(kmz_file, kml_file)
+
+    def download(self):
+        url = self.url_fires
+        answer = requests.get(url)
```
Comment (asmoiseenko) [inline import_kml.py:57]: rename to `response`. The answer is for the question.

File: import_kml.py [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+        for file in kml.filelist:
+            file_name = file.filename
+            if re.search('.kml$', file_name) is None:
+                continue
+            else:
+                if kml_filename is None:
+                    kml_filename = file_name
+                else:
+                    self.integration_log.add_log(LogLevel.ERROR, f'Failed extract kml file. Exception [kmz file stores more than one kml file]')
+                    raise Exception(f'Failed extract kml file. Exception [kmz file stores more than one kml file]')
```
Comment (asmoiseenko) [inline import_kml.py:86]: I see nothing about this limitation in the readme. Also, you didn't address the following comment https://github.com/ov-integrations/kml_import/pull/1#discussion_r679360947

File: import_kml.py [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+
+    def download(self):
+        url = self.url_fires
+        answer = requests.get(url)
+        if answer.ok:
+            self.integration_log.add_log(LogLevel.INFO, 'KMZ file downloaded')
+            return answer.content
+        else:
+            self.integration_log.add_log(LogLevel.ERROR, f'Failed download kmz. Exception [{answer.text}]')
+            raise Exception(f'Failed download kmz. Exception [{answer.text}]')
```
Comment (asmoiseenko) [inline import_kml.py:63]: Why don`t you have a common global exception handling block where you will log errors?

File: import_kml.py [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+                self.integration_log.add_log(LogLevel.ERROR, f'Failed parse kml. Exception [Field fire_size not found]')
+                raise Exception(f'Failed parse kml. Exception [Field fire_size not found]')
+
+            coordinates = placemark.find('Point').find('coordinates')
+            if coordinates is None:
+                self.integration_log.add_log(LogLevel.ERROR, f'Failed parse kml. Exception [Field coordinates not found]')
+                raise Exception(f'Failed parse kml. Exception [Field coordinates not found]')
+
+            coordinates = re.split(',', coordinates.text)
+            parse_list.append({CSVHeader.REG_ID.value:registration_id, CSVHeader.NAME.value:name, CSVHeader.DESCRIPTION.value:description, \
```
Comment (asmoiseenko) [inline import_kml.py:157]: You have to document CSV file structure in the readme.

File: start_integration.py [hunk 20]
```diff
@@ -1,34 +1,42 @@
...
-import_name = password_data['importName']
+url_kmz = settings_data['urlKMZ']
+url_onevizion = settings_data['urlOneVizion']
+ov_access_key = settings_data['ovAccessKey']
+ov_secret_key = settings_data['ovSecretKey']
+import_name = settings_data['importName']
+import_action = settings_data['importAction']
 
-with open('ihub_parameters.json', "rb") as PFile:
+if re.search('https', settings_data['urlOneVizion']) is None:
```
Comment (asmoiseenko) [inline start_integration.py:26]: make your regexp work both with and without "s" in the URL protocol and remove this condition

**Review by asmoiseenko — CHANGES_REQUESTED** [review]
Also, add XML with a sample OneVizion config, so we can deploy and test this integration somewhere

File: import_kml.py (hunk 15 above)
Comment (P4radise) [reply import_kml.py:13]: Okay, I will replace the file names, but in any case the parser is designed for a specific kml file, so many variables use the name fire

File: import_kml.py (hunk 17 above)
Comment (P4radise) [reply import_kml.py:86]: Okay, I'll add the information to the readme.
I missed that comment. But I don't understand it. Do I need to add an example of a kml file structure that is supported by this integration?

File: start_integration.py (hunk 20 above)
Comment (P4radise) [reply start_integration.py]: Done

File: import_kml.py (hunk 18 above)
Comment (P4radise) [reply import_kml.py]: Done

File: import_kml.py (hunk 17 above)
Comment (P4radise) [reply import_kml.py]: I added the kml structure to the readme

File: import_kml.py (hunk 16 above)
Comment (P4radise) [reply import_kml.py]: Done

File: kmz_to_kml.py (hunk 9 above)
Comment (P4radise) [reply kmz_to_kml.py]: Got it, fixed it

**Comment by P4radise** [comment]
> Also, add XML with a sample OneVizion config, so we can deploy and test this integration somewhere

Done

File: kmz_to_kml.py (hunk 9 above)
Comment (glebowadim) [reply kmz_to_kml.py]: Oh, it's OK. Sorry. 
"Exception chaining happens automatically when an exception is raised inside an except or finally section."

**Review by glebowadim — CHANGES_REQUESTED** [review]

File: kmz_to_kml.py (hunk 9 above)
Comment (asmoiseenko) [reply kmz_to_kml.py]: you have to **handle**, **not collect** exceptions in a single place

File: import_kml.py [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+        self.import_action = import_action
+        self.integration_log = integration_log
+        self.integration_exception = integration_exception
+
+    def import_process(self, csv_file):
+        try:
+            import_id = self.get_import()
+            self.start_import(import_id, csv_file)
+        except Exception as e:
+            raise Exception(e)
```
[the remark this answers is not in the document]
Comment (P4radise) [reply import_kml.py]: I deleted it

File: import_kml.py [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+from enum import Enum
+import requests
+import zipfile
+import csv
+import re
+import os
+
+
+class Integration:
+    KMZ = 'file.kmz'
```
[the remark this answers is not in the document]
Comment (P4radise) [reply import_kml.py]: Done

File: import_kml.py [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+
+    def start_integration(self):
+        self.integration_log.add_log(LogLevel.INFO, 'Starting Integration')
+
+        try:
+            self.kml.get(Integration.KMZ, Integration.KML)
+            parse_list = self.csv.parse(Integration.KML)
+            self.csv.create(parse_list, Integration.CSV)
+            self.integration_import.import_process(Integration.CSV)
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (P4radise) [reply import_kml.py]: I deleted it

File: import_kml.py [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+
+
+class KML:
+
+    def __init__(self, url, integration_log, integration_exception):
+        self.url = url
+        self.integration_log = integration_log
+        self.integration_exception = integration_exception
+
+    def get(self, kmz_file, kml_file):
```
[the remark this answers is not in the document]
Comment (P4radise) [reply import_kml.py]: Done

File: kmz_to_kml.py (hunk 9 above)
Comment (P4radise) [reply kmz_to_kml.py]: I have already handled these exceptions by adding conditions

File: import_kml.py (hunk 15 above)
Comment (asmoiseenko) [reply import_kml.py]: think about how you could make our code more generic. You just parse the placemark description and save values from it to the dedicated field. You may allow users to specify regexp patterns to use for values extraction in the integration config, see how it was done in the ov-monitoring integration:

settings.json
```json
{
   "extractPlacemarkDescriptionValues": [
        {"fireSize": "\d+\sacres|\d+acres"}, 
        {"date":"\d+-\w+-\d+"}
    ]

}
```

File: import_kml.py [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+        else:
+            self.integration_exception.add(LogLevel.ERROR, f'Failed to receive import. Exception [{str(response.text)}]')
+
+        return import_id
+
+    def start_import(self, import_id, file_name):
+        OVImport(self.url_onevizion_without_protocol, self.access_key, self.secret_key, import_id, file_name, self.import_action, isTokenAuth=True)
+
+
+class IntegrationException:
```
[the remark this answers is not in the document]
Comment (P4radise) [reply import_kml.py]: Done

File: import_kml.py (hunk 15 above)
Comment (P4radise) [reply import_kml.py]: Okay, I see your point, but let's save that change for the next Issue.
Now let's run this integration for a particular kml structure.

File: start_integration.py [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+process_id = ihub_data['processId']
+log_level = ihub_data['logLevel']
+
+integration_log = IntegrationLog(process_id, url_onevizion_without_protocol, ov_access_key, ov_secret_key, None, True, log_level)
+kml_file = KML(url_kmz, integration_log)
+csv_file = CSV(integration_log)
+integration_import = Import(url_onevizion, url_onevizion_without_protocol, ov_access_key, ov_secret_key, import_name, import_action, integration_log)
+integration = Integration(kml_file, csv_file, integration_import, integration_log)
+
+integration.start_integration()
```
Comment (glebowadim) [inline start_integration.py:40]: Move exception handling here

File: import_kml.py [hunk 27]
```diff
@@ -0,0 +1,220 @@
...
+        parse_list = []
+        with open(kml_filename, 'r') as kml:
+            parse = BeautifulSoup(kml, 'lxml-xml')
+
+        description_with_date = parse.find('description')
+        if description_with_date is None:
+            raise Exception(f'Failed parse KML. Exception [Description for registration_id not found]')
+
+        date = re.search('\d+-\w+-\d+', description_with_date.text)
+        if date is None:
```
Comment (glebowadim) [inline import_kml.py:109]:
```  
if date is None:
    raise Exception(f'Failed parse KML. Exception [Date for registration_id not found]')

registration_id = f'Fires-{datetime.strptime(date.group(), "%d-%b-%Y").strftime(...
```

File: import_kml.py [hunk 28]
```diff
@@ -0,0 +1,220 @@
...
+            file_name = file.filename
+            if re.search('.kml$', file_name) is None:
+                continue
+            else:
+                if kml_filename_to_extract is None:
+                    kml_filename_to_extract = file_name
+                else:
+                    raise Exception(f'Failed extract KML file. Exception [KMZ file stores more than one KML file]')
+
+        if kml_filename_to_extract is None:
```
Comment (glebowadim) [inline import_kml.py:86]:
```  
if kml_filename_to_extract is None:
    raise Exception(f'Failed extract KML file. Exception [KML file not found in KMZ]')
kml.extract(kml_filename_to_extract)
os.rename(kml_filename_to_extract, kml_filename)
self.integration_log.add(LogLevel.INFO, 'KML file extracted')
```

File: import_kml.py [hunk 29]
```diff
@@ -0,0 +1,220 @@
...
+        url = f'{self.url_onevizion}/api/v3/imports'
+        response = requests.get(url, headers={'Content-type':'application/json', 'Content-Encoding':'utf-8', 'Authorization':f'Bearer {self.access_key}:{self.secret_key}'})
+        if response.ok:
+            for import_data in response.json():
+                import_name = import_data['name']
+                if import_name == self.import_name:
+                    import_id = import_data['id']
+                    break
+
+            if import_id is None:
```
Comment (glebowadim) [inline import_kml.py:197]:
```        
  if import_id is None:
      raise Exception(f'Import \"{self.import_name}\" not found')
  self.integration_log.add(LogLevel.INFO, f'Import \"{self.import_name}\" founded')
```

**Review by glebowadim — CHANGES_REQUESTED** [review]

File: import_kml.py (hunk 29 above)
Comment (P4radise) [reply import_kml.py:194]: Done

File: import_kml.py (hunk 28 above)
Comment (P4radise) [reply import_kml.py:83]: Done

File: import_kml.py (hunk 27 above)
Comment (P4radise) [reply import_kml.py:106]: Done

File: start_integration.py (hunk 26 above)
Comment (P4radise) [reply start_integration.py]: Done

**Review by glebowadim — APPROVED** [review]

**Pull request merged**